### PR TITLE
Allow configurable temp directory for query processing

### PR DESCRIPTION
**Pull request by jon-wei**
Adds a temp directory option to the group by query config 

Fixes https://github.com/druid-io/druid/issues/3776

### CODE REVIEW

File: docs/content/querying/groupbyquery.md [hunk 1]
```diff
@@ -182,6 +182,7 @@ When using the "v2" strategy, the following runtime properties apply:
 |`druid.query.groupBy.bufferGrouperMaxLoadFactor`|Maximum load factor of the off-heap hash table used for grouping results. When the load factor exceeds this size, the table will be grown or spilled to disk. Set to 0 to use a reasonable default.|0|
 |`druid.query.groupBy.maxMergingDictionarySize`|Maximum amount of heap space (approximately) to use for the string dictionary during merging. When the dictionary exceeds this size, a spill to disk will be triggered.|100000000|
 |`druid.query.groupBy.maxOnDiskStorage`|Maximum amount of disk space to use, per-query, for spilling result sets to disk when either the merging buffer or the dictionary fills up. Queries that exceed this limit will fail. Set to zero to disable disk spilling.|0 (disabled)|
+|`druid.query.groupBy.tmpDir`|Path where temporary files created while processing the query should be stored. If specified, this configuration takes priority over the default `java.io.tmpdir` path.|
```
Comment (gianm) [inline docs/content/querying/groupbyquery.md]: How about putting this in DruidProcessingConfig? It feels more like something that should apply to any query that needs a temp scratch area. Kind of senseless for each query to manage its own scratch areas.

File: docs/content/querying/groupbyquery.md [hunk 2]
```diff
@@ -214,3 +215,4 @@ When using the "v2" strategy, the following query context parameters apply:
 |`maxMergingDictionarySize`|Can be used to lower the value of `druid.query.groupBy.maxMergingDictionarySize` for this query.|
 |`maxOnDiskStorage`|Can be used to lower the value of `druid.query.groupBy.maxOnDiskStorage` for this query.|
 |`sortByDimsFirst`|Sort the results first by dimension values and then by timestamp.|
+|`tmpDir`|Overrides the value of `druid.query.groupBy.tmpDir` for this query.|
```
Comment (gianm) [inline docs/content/querying/groupbyquery.md]: I'd leave this out. I can't think of a good reason for letting a user override the temp dir used on Druid servers, and this is a security hole (user could set tmpDir = somewhere the user shouldn't be able to scribble on).

File: docs/content/querying/groupbyquery.md (hunk 1 above)
Comment (gianm) [inline docs/content/querying/groupbyquery.md]: There should be another column with the default.

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java [hunk 3]
```diff
@@ -129,8 +129,9 @@ public GroupByMergingQueryRunnerV2(
       combiningAggregatorFactories[i] = query.getAggregatorSpecs().get(i).getCombiningFactory();
     }
 
+    final String queryTmpDir = querySpecificConfig.getTmpDir();
     final File temporaryStorageDirectory = new File(
-        System.getProperty("java.io.tmpdir"),
+        queryTmpDir != null ? queryTmpDir : System.getProperty("java.io.tmpdir"),
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java]: Would be better to have this default populated in the config object's "getTmpDir" method, so any other engines that need scratch space would automatically agree on the default location.

**Comment by jon-wei** [comment]
@gianm updated this to place the temp dir config in `druid.processing.tmpDir` instead

**Comment by jihoonson** [comment]
Looks good to me.

**Pull request merged**